packages/@aws-cdk/aws-appmesh/lib/tls-certificate.ts

```diff
@@ -0,0 +1,173 @@
+   */
+  STRICT = 'STRICT',
+
+  /**
+   * Accept encrypted and plaintext traffic.
+   */
+  PERMISSIVE = 'PERMISSIVE',
+
+  /**
+   * TLS is disabled, only accept plaintext traffic.
+   */
+  DISABLED = 'DISABLED',
+}
+
+/**
+ * A wrapper for the tls config returned by {@link TlsCertificate.bind}
+ */
+export interface TlsCertificateConfig {
+  /**
+   * The CFN shape for a listener TLS certificate
+   */
+  readonly tlsCertificate: CfnVirtualNode.ListenerTlsCertificateProperty,
+
+  /**
+   * The TLS mode.
+   */
+  readonly tlsMode: TlsMode;
+}
+
+/**
+ * ACM Certificate Properties
+ */
+export interface AcmCertificateOptions {
+  /**
+   * The TLS mode.
+   */
+  readonly tlsMode: TlsMode;
+
+  /**
+   * The ACM certificate
+   */
+  readonly certificate: acm.ICertificate;
+}
+
+/**
+ * File Certificate Properties
+ */
+export interface FileCertificateOptions {
+  /**
+   * The TLS mode.
+   */
+  readonly tlsMode: TlsMode;
+
+  /**
+   * The file path of the certificate chain file.
+   */
+  readonly certificateChainPath: string;
+
+  /**
+   * The file path of the private key file.
+   */
+  readonly privateKeyPath: string;
+}
+
+/**
+ * Represents a TLS certificate
+ */
+export abstract class TlsCertificate {
+  /**
+   * Returns an File TLS Certificate
+   */
+  public static file(props: FileCertificateOptions): TlsCertificate {
+    return new FileTlsCertificate(props);
+  }
+
+  /**
+   * Returns an ACM TLS Certificate
+   */
+  public static acm(props: AcmCertificateOptions): TlsCertificate {
+    return new AcmTlsCertificate(props);
+  }
+
+  /**
+   * Returns TLS certificate based provider.
+   */
+  public abstract bind(_scope: cdk.Construct): TlsCertificateConfig;
+
+}
+
+/**
+ * Represents a ACM provided TLS certificate
+ */
+class AcmTlsCertificate extends TlsCertificate {
+  /**
+   * The TLS mode.
+   *
+   * @default - TlsMode.DISABLED
+   */
+  readonly tlsMode: TlsMode;
+
+  /**
+   * The ARN of the ACM certificate
+   */
+  readonly acmCertificate: acm.ICertificate;
+
+  constructor(props: AcmCertificateOptions) {
+    super();
+    this.tlsMode = props.tlsMode ? props.tlsMode : TlsMode.DISABLED;
+    this.acmCertificate = props.certificate;
+  }
+
+  bind(_scope: cdk.Construct): TlsCertificateConfig {
+    return {
+      tlsCertificate: {
+        acm: {
+          certificateArn: this.acmCertificate.certificateArn,
+        },
+      },
+      tlsMode: this.tlsMode,
+    };
+  }
+}
+
+/**
+ * Represents a file provided TLS certificate
+ */
+class FileTlsCertificate extends TlsCertificate {
+  /**
+   * The TLS mode.
+   *
+   * @default - TlsMode.DISABLED
+   */
+  readonly tlsMode: TlsMode;
+
+  /**
+   * The file path of the certificate chain file.
+   */
+  readonly certificateChain: string;
+
+  /**
+   * The file path of the private key file.
+   */
+  readonly privateKey: string;
+
+  constructor(props: FileCertificateOptions) {
+    super();
+    this.tlsMode = props.tlsMode ? props.tlsMode : TlsMode.DISABLED;
```

Comment:
If this is a required property, there is no reason to do a check here.

Reply:
ah thanks I should've removed this. I had tlsMode at the listener level in a previous revision.